README.md

```diff
@@ -0,0 +1,121 @@
+
+This prototype is specific to generically managing actual Kubernetes config, it
+does not wrap Kubernetes yaml in Ansible yaml, or expose individual modules per
+individual Kubernetes type.
+
+This prototype aims to make applying Kubernetes config via Ansible extremely
+easy, including on-the-fly templating and simple yaml edits, file based or
+inline config, and patching.
+
+`kubectl delete` is represented by state: absent on the kubectl module itself.
+
+`kubectl get` is represented by a kubectl_facts module.
+
+## Assumptions
+
+  * kubectl apply will be modified to accurately [report if changes had to be applied](https://github.com/kubernetes/kubernetes/pull/52647).
+
+# Target Ansible
+
+```yaml
+- kubectl:
+
+    # State present implies kubectl apply, and absent implies delete. Many options
+    # below would not be relevant for delete.
+    state: present
+
+    kubeconfig:
+      # Kubeconfig file on the remote host:
+      file: /etc/origin/master/admin.kubeconfig
+      # Alternatively:
+      inline: "{{ lookup('file', 'local_on_control_host.kubeconfig') }}"
+
+    # Optional, omit if your configuration specifies a namespace, or you're operating on
+    # cluster level objects.
+    namespace: default
+
+    # You can specify your kubernetes config inline if desired. The | is important here:
+    inline: |
+      kind: "Namespace"
+      apiVersion: v1
+      metadata:
+        name: testnamespace
+
+    # Alternatively you can use files that live with your role/playbook. This would be
+    # mutually exclusive with inline data.
+    # Each file is copied over to the remote host, potentially edited or patched, and
+    # then submitted with apply.
+    # Specifying multiple files implies one execution of kubectl apply. If you want multiple
+    # you can use separate tasks or a with_items loop.
+    files:
+    - src: configmap.json
+    - src: subdir/
+    - src: service.yml
+      # Support small edits to the yaml on the fly, without going to a full template. Operates
+      # on the temporary copy of the file sent to the remote host.
+      yedit:
+        key: spec.metadata.name
+        value: "{{ replacement_name }}"
+        value_type: list
+    - src: patchme.yml
+      # Support supplying structured patches with "kubectl patch" prior to submitting to the server.
+      patches:
+      - patch1.yml
+      - patch2.yml
+    # This would be nice too, kubectl supports, would likely imply no yediting or patching unless
+    # we want the module to be in the business of fetching files.
+    - src: https://github.com/me/project/template.yml
+
+    # Support templating files out automatically. Use Ansible's native mechanisms for jinja
+    # templates but don't require the user to copy the file and manually clean it up with
+    # their own tasks.
+    templates:
+    - src: apptemplate.yml
+      vars:
+        foo: bar
```

Comment:
In theory we could allow yedit/patch on the templates as well but may not be necessary.

Reply:
I am curious if Ansible core is open to a module with templating embedded, or if this is not acceptable and users would be expected to manually use the template module, and cleanup files themselves. As a user, I'd love to see this handled for me.